src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +201,83 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl(kj::mv(precommitAlarmState));
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::requestScheduledAlarm(kj::Maybe<kj::Date> requestedTime) {
+  // Not using coroutines here, because it's important for correctness in workerd that a
+  // synchronously thrown exception in scheduleRun() can escape synchronously to the caller.
+  return hooks.scheduleRun(requestedTime).then([this, requestedTime]() {
+    lastConfirmedScheduledAlarm = requestedTime;
+  });
+}
+
+ActorSqlite::PrecommitAlarmState ActorSqlite::startPrecommitAlarmScheduling() {
+  PrecommitAlarmState state;
+  state.localAlarmState = metadata.getAlarm();
+  if (pendingCommit == kj::none &&
+      willFireEarlier(state.localAlarmState, lastConfirmedScheduledAlarm)) {
+    // Basically, this is the first scheduling request that commitImpl() would make prior to
+    // commitCallback().  We start the request separately, ahead of calling sqlite functions that
+    // commit to local disk, for correctness in workerd, where alarm scheduling and db commits are
+    // both synchronous.
+    state.schedulingPromise = requestScheduledAlarm(state.localAlarmState);
+  }
+  return kj::mv(state);
+}
+
+kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState precommitAlarmState) {
+  // We assume that exceptions thrown during commit will propagate to the caller, such that they
+  // will ensure cancelDeferredAlarmDeletion() is called, if necessary.
+
+  KJ_IF_SOME(pending, pendingCommit) {
+    // If an earlier commitImpl() invocation is already in the process of updating precommit
+    // alarms but has not yet made the commitCallback() call, it should be OK to wait on it to
+    // perform the precommit alarm update and db commit for this invocation, too.
+    co_await pending.addBranch();
+    co_return;
+  }
+  auto [promise, fulfiller] = kj::newPromiseAndFulfiller<void>();
+  pendingCommit = promise.fork();
+
+  // Wait for the first precommit alarm scheduling request to complete, if any.
+  auto localAlarmState = precommitAlarmState.localAlarmState;
+  KJ_IF_SOME(p, precommitAlarmState.schedulingPromise) {
+    co_await p;
```

Comment:
If it throws, I believe it should cause the coroutine context to be destroyed, which should destroy `fulfiller`, which should cause pendingCommit and all other forked promises to reject.  Kenton had the same concern, so maybe I'm misunderstanding something, but I did add a test case that I think would fail if a thrown exception caused hangs:

https://github.com/cloudflare/workerd/blob/bee6326be6a8f378cee2f5086526fd178e2e3f49/src/workerd/io/actor-sqlite-test.c%2B%2B#L826-L859

Reply:
You're right, I forgot about the auto-reject-on-destruct behavior of fulfillers. Thanks for explaining!

Reply:
Added comment in d1983124b .